crates/api/src/local_user.rs

```diff
@@ -180,58 +187,83 @@ impl Perform for Register {
       preferred_username: None,
       published: None,
       updated: None,
-      admin: no_admins,
-      banned: Some(false),
-      show_nsfw: data.show_nsfw,
-      theme: "browser".into(),
-      default_sort_type: SortType::Active as i16,
-      default_listing_type: ListingType::Subscribed as i16,
-      lang: "browser".into(),
-      show_avatars: true,
-      send_notifications_to_email: false,
-      actor_id: Some(user_actor_id.clone()),
+      banned: None,
+      deleted: None,
+      actor_id: Some(actor_id.clone()),
       bio: None,
-      local: true,
-      private_key: Some(user_keypair.private_key),
-      public_key: Some(user_keypair.public_key),
+      local: Some(true),
+      private_key: Some(Some(actor_keypair.private_key)),
+      public_key: Some(Some(actor_keypair.public_key)),
       last_refreshed_at: None,
-      inbox_url: Some(generate_inbox_url(&user_actor_id)?),
-      shared_inbox_url: Some(Some(generate_shared_inbox_url(&user_actor_id)?)),
+      inbox_url: Some(generate_inbox_url(&actor_id)?),
+      shared_inbox_url: Some(Some(generate_shared_inbox_url(&actor_id)?)),
+    };
+
+    // insert the person
+    let inserted_person = match blocking(context.pool(), move |conn| {
+      Person::create(conn, &person_form)
+    })
+    .await?
+    {
+      Ok(u) => u,
+      Err(_) => {
+        return Err(ApiError::err("user_already_exists").into());
+      }
+    };
+
+    // Create the local user
+    let local_user_form = LocalUserForm {
+      person_id: inserted_person.id,
+      email: Some(data.email.to_owned()),
+      matrix_user_id: None,
+      password_encrypted: data.password.to_owned(),
+      admin: Some(no_admins),
+      show_nsfw: Some(data.show_nsfw),
+      theme: Some("browser".into()),
+      default_sort_type: Some(SortType::Active as i16),
+      default_listing_type: Some(ListingType::Subscribed as i16),
+      lang: Some("browser".into()),
+      show_avatars: Some(true),
+      send_notifications_to_email: Some(false),
     };
 
-    // Create the user
-    let inserted_user = match blocking(context.pool(), move |conn| {
-      User_::register(conn, &user_form)
+    let inserted_local_user = match blocking(context.pool(), move |conn| {
+      LocalUser::register(conn, &local_user_form)
     })
     .await?
     {
-      Ok(user) => user,
+      Ok(lu) => lu,
       Err(e) => {
         let err_type = if e.to_string()
-          == "duplicate key value violates unique constraint \"user__email_key\""
+          == "duplicate key value violates unique constraint \"local_user_email_key\""
         {
           "email_already_exists"
         } else {
           "user_already_exists"
         };
 
+        // If the local user creation errored, then delete that person
+        blocking(context.pool(), move |conn| {
+          Person::delete(&conn, inserted_person.id)
+        })
+        .await??;
+
         return Err(ApiError::err(err_type).into());
```

Comment:
Annoying but necessary, when a new user registers, we have to insert rows in both the `person` and `local_user` table. I'm doing a check here to see if there was an error inserting into the `local_user`, and if so, deleting that `inserted_person`.

Reply:
Why dont you insert the local_user first? If that works without errors, inserting the Person should be guaranteed to work (I think).

Reply:
local user table has a foreign key to person, it requires a `person_id`